core/engine/src/builtins/number/globals.rs

```diff
@@ -297,40 +303,71 @@ pub(crate) fn parse_float(
-        } else if s.starts_with("-Infinity") {
-            Ok(JsValue::new(f64::NEG_INFINITY))
-        } else if s_prefix_lower.starts_with("inf")
-            || s_prefix_lower.starts_with("+inf")
-            || s_prefix_lower.starts_with("-inf")
-        {
-            // Prevent fast_float from parsing "inf", "+inf" as Infinity and "-inf" as -Infinity
-            Ok(JsValue::nan())
-        } else {
-            Ok(fast_float2::parse_partial::<f64, _>(s).map_or_else(
-                |_| JsValue::nan(),
-                |(f, len)| {
-                    if len > 0 {
-                        JsValue::new(f)
-                    } else {
-                        JsValue::nan()
-                    }
-                },
-            ))
+    const PLUS_CHAR: u16 = b'+' as u16;
+    const MINUS_CHAR: u16 = b'-' as u16;
+    const LOWER_CASE_I_CHAR: u16 = b'i' as u16;
+    const UPPER_CASE_I_CHAR: u16 = b'I' as u16;
+
+    let Some(string) = args.first() else {
+        return Ok(JsValue::nan());
+    };
+
+    // OPTIMIZATION: We can skip the round-trip when the value is already a number.
+    if string.is_number() {
+        // Special case for negative zero - it should become positive zero
+        if string.is_negative_zero() {
+            return Ok(JsValue::new(0));
         }
-    } else {
-        // Not enough arguments to parseFloat.
-        Ok(JsValue::nan())
+
+        return Ok(string.clone());
+    }
+
+    // 1. Let inputString be ? ToString(string).
+    let input_string = string.to_string(context)?;
+
+    // 2. Let trimmedString be ! TrimString(inputString, start).
+    let trimmed_string = input_string.trim_start();
+
+    // 3. Let trimmed be StringToCodePoints(trimmedString).
+    // 4. Let trimmedPrefix be the longest prefix of trimmed that satisfies the syntax of a StrDecimalLiteral, which might be trimmed itself. If there is no such prefix, return NaN.
+    // 5. Let parsedNumber be ParseText(trimmedPrefix, StrDecimalLiteral).
+    // 6. Assert: parsedNumber is a Parse Node.
+    // 7. Return the StringNumericValue of parsedNumber.
+    let (positive, prefix) = match trimmed_string.get(0) {
+        Some(PLUS_CHAR) => (true, trimmed_string.get(1..).unwrap_or(JsStr::latin1(&[]))),
+        Some(MINUS_CHAR) => (false, trimmed_string.get(1..).unwrap_or(JsStr::latin1(&[]))),
+        _ => (true, trimmed_string),
+    };
+
+    if prefix.starts_with(js_str!("Infinity")) {
+        if positive {
+            return Ok(JsValue::positive_infinity());
+        }
+        return Ok(JsValue::negative_infinity());
+    } else if let Some(LOWER_CASE_I_CHAR | UPPER_CASE_I_CHAR) = prefix.get(0) {
+        return Ok(JsValue::nan());
     }
+
+    let value = match trimmed_string.variant() {
+        JsStrVariant::Latin1(s) => fast_float2::parse_partial::<f64, _>(s),
+        JsStrVariant::Utf16(s) => {
+            let s = String::from_utf16_lossy(s);
+            fast_float2::parse_partial::<f64, _>(s.as_bytes())
```

Comment:
thought: it would be nice if `fast_float2` could parse directly from `UTF16`.

Maybe add a `NOTE` comment as a potential future optimization.